File: JavaPractice/src/BinaryTreeCreateLinkedListEachLevel.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class BinaryTreeCreateLinkedListEachLevel {

	public static class TreeNode {
		int data;
		TreeNode right;
		TreeNode left;

		public TreeNode(int d) {
			this.data = d;
			this.right = null;
			this.left = null;
		}

	}

	public static List<LinkedList<TreeNode>> createList(TreeNode root) {

		List<LinkedList<TreeNode>> lists = new ArrayList<LinkedList<TreeNode>>();
		createList(root, lists, 0);
		return lists;
	}

	public static void createList(TreeNode root, List<LinkedList<TreeNode>> lists, int level) {
		if (root == null)
			return;

		LinkedList<TreeNode> list = null;
		if (lists.size() == level) {
			list = new LinkedList<TreeNode>();
			lists.add(list);
		} else {
			list = lists.get(level);
		}
		list.add(root);
		createList(root.left, lists, level + 1);
		createList(root.right, lists, level + 1);
	}

	public static void display(List<LinkedList<TreeNode>> root) {
		if (root != null) {
			for (LinkedList<TreeNode> l : root) {
				for (TreeNode r : l) {
					System.out.print(r.data + "\t");
				}
				System.out.println();
			}
		}
	}

	public static void main(String arg[]) {

		TreeNode node = new TreeNode(5);
		node.right = new TreeNode(8);
		node.left = new TreeNode(4);
		node.right.left = new TreeNode(13);
		node.left.left = new TreeNode(11);
		node.left.left.right = new TreeNode(2);
		node.right.right = new TreeNode(4);
		node.right.right.right = new TreeNode(1);
		node.left.left.left = new TreeNode(7);
		node.right.right.left = new TreeNode(5);
		List<LinkedList<TreeNode>> s = createList(node);
		display(s);
	}

}

File: PythonPractice/src/BinaryTreeCreateLinkedListEachLevel.py
from typing import Optional, List, Deque
from collections import deque

class TreeNode:
    def __init__(self, data: int):
        if not isinstance(data, int):
            raise TypeError("Data must be an integer")

        self.data: int = data
        self.left: Optional['TreeNode'] = None
        self.right: Optional['TreeNode'] = None

def create_list(root: Optional[TreeNode]) -> List[Deque[TreeNode]]:
    if not isinstance(root, (TreeNode, type(None))):
        raise TypeError("Root must be a TreeNode or None")

    lists: List[Deque[TreeNode]] = []
    _create_list(root, lists, 0)
    return lists

def _create_list(root: Optional[TreeNode], lists: List[Deque[TreeNode]], level: int):
    if root is None:
        return

    if len(lists) == level:
        lists.append(deque())

    lists[level].append(root)

    _create_list(root.left, lists, level + 1)
    _create_list(root.right, lists, level + 1)

def display(lists: List[Deque[TreeNode]]):
    if not isinstance(lists, list):
        raise TypeError("Lists must be of type List")

    for level in lists:
        if not isinstance(level, Deque):
            raise TypeError("Each level must be a Deque")

        print("\t".join(str(node.data) for node in level))
        print()

if __name__ == "__main__":
    # Create the binary tree
    node = TreeNode(5)
    node.right = TreeNode(8)
    node.left = TreeNode(4)
    node.right.left = TreeNode(13)
    node.left.left = TreeNode(11)
    node.left.left.right = TreeNode(2)
    node.right.right = TreeNode(4)
    node.right.right.right = TreeNode(1)
    node.left.left.left = TreeNode(7)
    node.right.right.left = TreeNode(5)

    s = create_list(node)
    display(s)
